File: Single valued subtree.py
#User function Template for python3

'''
class Node:
    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None
'''

class Solution:
    def countSingleRec(self,root , count):
        #code here
        if root is None :
            return True
 
        # Recursively count in left and right subtress also
        left = self.countSingleRec(root.left , count)
        right = self.countSingleRec(root.right , count)
     
        # If any of the subtress is not singly, then this
        # cannot be singly
        if left == False or right  == False :
            return False
     
        # If left subtree is singly and non-empty , but data
        # doesn't match
        if root.left and root.data != root.left.data:
            return False
 
        # same for right subtree
        if root.right and root.data != root.right.data:
            return False
 
        # If none of the above conditions is True, then
        # tree rooted under root is single valued,increment
        # count and return true
        count[0] += 1
        return True
    
    def singlevalued(self, root):
        #code here
        count = [0]
 
        # Recursive function to count
        self.countSingleRec(root , count)
 
        return count[0]
        

File: Single valued subtree.java
class Count
{
    int count = 0;
}
class Solution
{
    Node root; 
    Count ct = new Count();
    boolean countSingleRec(Node node, Count c)
    {
        // Return false to indicate NULL
        if (node == null)
            return true;
          
        // Recursively count in left and right subtrees also
        boolean left = countSingleRec(node.left, c);
        boolean right = countSingleRec(node.right, c);
  
        // If any of the subtrees is not singly, then this
        // cannot be singly.
        if (left == false || right == false)
            return false;
  
        // If left subtree is singly and non-empty, but data
        // doesn't match
        if (node.left != null && node.data != node.left.data)
            return false;
  
        // Same for right subtree
        if (node.right != null && node.data != node.right.data)
            return false;
  
        // If none of the above conditions is true, then
        // tree rooted under root is single valued, increment
        // count and return true.
        c.count++;
        return true;
    }
    int singlevalued()
    {
        return singlevalued(root);
    }
  
    public int singlevalued(Node root)
    {
        //code here
        countSingleRec(root, ct);
        return ct.count;
    }
}
